File: sicp/procedures/basic/square.hs

sqrtx x = sqrt_iter 1 x
	where
		sqrt_iter::(Floating a, Ord a) => a -> a -> a
		sqrt_iter guess x =
			if (good_enough) then 
				guess
			else
				sqrt_iter improve x
			where
				good_enough  	= (abs ((square guess) - x)) < 0.001
				improve  		= average guess (x / guess)
				square x 		= x * x
				average x y 	= (x + y) / 2

main = do
		print $ sqrtx (fromIntegral 10)

File: old/sicp/procedures/basic/square.scala

object sqrttest {

	import scala.math._
	import scala.annotation._
	def sqrt(x:Double) = {
		var guess:Double = 1

		def good_enough = abs(square(guess) - x) < 0.001
		def improve = guess = average(guess, x / guess)
		def average(x:Double, y:Double) = (x + y) / 2
		def square(x:Double) = x * x

		@tailrec
		def sqrt_iter():Double = 
			if (good_enough) guess else { improve ; sqrt_iter}

		sqrt_iter	
	}

	def main(args:Array[String]) {
		println(sqrt(10))
	}
}
